### PR TITLE
Feature/fhirpath date operations

### PR DESCRIPTION
**Pull request by brianpos**
This also required a change to the DynamicGet function too, which now makes it perform an exact match on types (in the same way that the normal casting version does) and if that doesn't find one, then uses the existing approach.

Using the linq expressions (and not putting them into lists) it will bail from the searching when it finds the first exact match, and won't even start looking for castable matches...

### CODE REVIEW

**Comment by brianpos** [comment]
@marcovisserFurore any update on which release this might go in, or if there are changes needed before merging in?

**Comment by ewoutkramer** [comment]
I think this will make it into the November release, @marcovisserFurore - will we have one?

**Comment by marcovisserFurore** [comment]
That would be a small release, but yes we can do that!

File: src/Hl7.FhirPath/FhirPath/Expressions/Typecasts.cs
Comment (ewoutkramer) [inline src/Hl7.FhirPath/FhirPath/Expressions/Typecasts.cs:148]: Shall we call this `IsOfExactType` - and maybe keep it `internal` for now....I don't like the duplication with `CanCastTo`, but we don't have to fix that right now.

**Review by ewoutkramer — CHANGES_REQUESTED** [review]

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:146]: I'll check where I found that from it was ucum related and from other code in the apis.
And also recheck I got weeks etc right.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:141]: Null values wouldn't pickup the type s as quantity or date though from the symbol table?

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:151]: I think you've read that wrong, I just rechecked it and think it's doing the right thing.
Adding 25 months will add 2 years and 1 month how it is now.
(I will follow up with the use of the a unit though)

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:156]: I think you've misread that part, but will add week.
The notes in the spec state that there are calendar units.

File: src/Hl7.FhirPath.Tests/Tests/BasicFunctionTests.cs
Comment (brianpos) [reply src/Hl7.FhirPath.Tests/Tests/BasicFunctionTests.cs:192]: It does. Being in the +10/+11 zone.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:151]: > For Date values, the quantity unit must be one of: years, months, weeks, or days

There is no "mo" unit there.... it's not a calender unit. And "mo" != "month"!

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:141]: That's right, but you can call the API separate from the symboltable, it's just a public function, so we should guard for it!

File: src/Hl7.FhirPath/FhirPath/Expressions/SymbolTable.cs
Comment (brianpos) [reply src/Hl7.FhirPath/FhirPath/Expressions/SymbolTable.cs:110]: I tested it and due to the enumeration yield if the exact match returns the single result, then it doesn't start enumerating into the other one.

File: src/Hl7.FhirPath/FhirPath/Expressions/Typecasts.cs
Comment (brianpos) [reply src/Hl7.FhirPath/FhirPath/Expressions/Typecasts.cs:148]: Happy to rename it.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:156]: 6.7 does not mention these UCUM units at all.  It actually says:

> To avoid the potential confusion of calendar-based date/time arithmetic with definite duration date/time arithmetic, FHIRPath defines definite-duration date/time arithmetic for seconds and below, and calendar-based date/time arithmetic for seconds and above. At the second, calendar-based and definite-duration-based date/time arithmetic are identical. **If a definite-quantity duration above seconds appears in a date/time arithmetic calculation, the evaluation will end and signal an error to the calling environment.**

Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:156]: The spec has explicitly described that '1 year' is not equal to '1 a', although it is equivalent `'1 year' ~ '1 a'`, but for the date/time calculations, it seems to explicitly forbid the use.

File: src/Hl7.FhirPath/FhirPath/Expressions/SymbolTable.cs
Comment (ewoutkramer) [reply src/Hl7.FhirPath/FhirPath/Expressions/SymbolTable.cs:110]: ahhh, right. great!

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:151]: Ah, sorry, I'll find why I added in a and mo.
Your statement wasn't about the evaluation. It was the unit literal.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:156]: I'll find why I put it in and get pack to you.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:141]: ArgumentNullException ok instead of invalid argument?
(same as in the tryparse)

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:151]: Any thoughts on how we can deal with this?
http://hl7.org/fhir/fhirpath.html#quantity
This is where it has explicitly states that 1a implicitly converts to 1 year

Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:151]: I guess we could put that one into the actual FHIR Quantity type?

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:141]: Nowadays you'll get a warning from the compiler if the name of the argument mentioned in the ArgumentNullException (value) is not literally "value" (the name of the parameter).  In this case `value` is not null, but `value.Value` is.  So, although I would fully agree with you, you'll get a slightly confusing NullException on "value", while "value" itself is not null.  But this is only recently, so I only switched to InvalidArgument recently.  Any other opinions, @marcovisserFurore ?

File: src/Hl7.FhirPath.Tests/Functions/OperationsTests.cs
Comment (brianpos) [reply src/Hl7.FhirPath.Tests/Functions/OperationsTests.cs:31]: Interesting, the 365 test doesn't pass there. Which means the .net DateTimeOffset class does it differently?

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:141]: Check my PR update, there's another property on the exception for the name of the property (and I corrected the property names from me and value to be more sensible)

Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:141]: Ah, I also see you referred to value.value, that would now be caught in the default units check, and thus get the appropriate message there, just argument exception.
Which now makes me think I need to recheck dateValue.value is that even possible?

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:159]: As noted above, that's in the fhir spec fhirpath page, not fhirpath itself.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:170]: This is now corrected.

File: src/Hl7.FhirPath/FhirPath/Expressions/Typecasts.cs
Comment (brianpos) [reply src/Hl7.FhirPath/FhirPath/Expressions/Typecasts.cs]: Done.

File: src/Hl7.FhirPath.Tests/Functions/OperationsTests.cs
Comment (brianpos) [reply src/Hl7.FhirPath.Tests/Functions/OperationsTests.cs:31]: https://www.timeanddate.com/date/dateadded.html?d1=01&m1=01&y1=2016&type=add&ay=&am=&aw=&ad=365&rec=
This isn't making sense to me why it should move up to the next year.
Where is the standard year defined? and is there a standard month defined too?

Comment (ewoutkramer) [reply src/Hl7.FhirPath.Tests/Functions/OperationsTests.cs:31]: The documentation for this testcase says:

"The first example above returns @2015, because 23 months only constitutes one year. The second example returns 2017 because even though 2016 is a leap-year, the time-valued quantity (365 days) is converted to 1 year, a standard calendar year of 365 days."

Comment (brianpos) [reply src/Hl7.FhirPath.Tests/Functions/OperationsTests.cs:31]: Have asked on zulip

Comment (ewoutkramer) [reply src/Hl7.FhirPath.Tests/Functions/OperationsTests.cs:31]: This is because of the way Grahame (but probably Bryn in the CQL spec) has described datetime operations:

> For partial date/time values where the time-valued quantity is more precise than the partial date/time, the operation is performed by converting the time-valued quantity to the highest precision in the partial (removing any decimal value off) and then adding to the date/time value.

Comment (ewoutkramer) [reply src/Hl7.FhirPath.Tests/Functions/OperationsTests.cs:31]: This for sure does not align with what the operations on DateTimeOffset do.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:141]: Recheck in what sense?

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:154]: We don't have an official "mapping", it's built into the FhirPath engine.  But thanks for pointing that out again, it reminds me that we need to do this in two steps:

* The "System/CQL" based types should *not* use "a" (as I said above).
* The "binding" between the FHIR poco/ITypedElement and the FhirPath engine should do the conversion Brian referred to. But we don't have that binding explicitly now.  We do could (for now) have an internal solution where for these new datetime additions (and maybe the comparisons/equality) we convert the units just before calling the functions in the symboltable?

File: src/Hl7.FhirPath.Tests/Functions/OperationsTests.cs
Comment (brianpos) [reply src/Hl7.FhirPath.Tests/Functions/OperationsTests.cs:31]: Happy with the explanation on zulip, and see the full table. Will update the implementation to cover that. (also reported it for the fhirpathjs implementation)

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:154]: The spec talks about to Quantity for ucum, but then when using on dates does those things. So I'm confused how we can know when to convert amd when not to.
Isn't it simpler just to handle it as I've put in?

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:159]: @ewoutkramer I've done the day/month converstions as they are documented in the spec, but there's no "conversion" for the weeks, should I treat it as though it was days?

Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:159]: Yes, let's do that.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:154]: Simpler, yes, but not really correct - FhirPath doesn't formally know about "a".  I had the same problem when dealing with comparisons.  Can you elaborate a bit on "when using on dates does those things" and "how we can know when to convert amd when not to"?

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (ewoutkramer) [inline src/Hl7.Fhir.ElementModel/Types/Date.cs:152]: Wouldn't this crash when `addValue.Value is null`?

Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:152]: this is the thing I mentioned above that I need to check and prevent in the Null Argument checking that I missed.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:162]: How would this work for you?
```
       case "weeks":
           if (dateValue.Precision == DateTimePrecision.Year)
                dto = dateValue._parsedValue.AddYears((int)(addValue.Value / 52));
           else if (dateValue.Precision == DateTimePrecision.Month)
                dto = dateValue._parsedValue.AddMonths((int)(addValue.Value * 7 / 30));
            else
                dto = dateValue._parsedValue.AddDays(((int)addValue.Value) * 7);
            break;
```
52 weeks in a year, and when at month level, assume was talking about average days in a month, or should it be more like `AddMonths((int)(addValue.Value/52*12))` so double based multiplication and division, then the cast trims the rounding to the values requested precision.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:152]: addValue.value is a decimal, not an object or nullable decimal, so can't be null.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:141]: This is fine, the addValue.value is a decimal not nullable, hence can't be null.
Believe is now safe.

File: src/Hl7.FhirPath.Tests/Functions/OperationsTests.cs
Comment (brianpos) [reply src/Hl7.FhirPath.Tests/Functions/OperationsTests.cs:31]: Now done.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:141]: Yep.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:154]: What I meant is that anywhere we are using a FHIR "Quantity" in the FhirPath engine and we expect a System.Quantity, we currently use the hard mapping built into Typecasts.cs:

```csharp
        internal static P.Quantity ParseQuantity(ITypedElement qe)
        {
            var value = qe.Children("value").SingleOrDefault()?.Value as decimal?;
            if (value == null) return null;

            var unit = qe.Children("code").SingleOrDefault()?.Value as string;
            return new P.Quantity(value.Value, unit);
        }
```

We could easily do the "a" -> "year" mapping there.  In fact, that would be the most natural place. It reflects that in FHIR the "a" UCUM unit is really (and erroneously, in retrospect) meant to be the calendar unit "year". And converting that unit here steers the core of the FP engine away from "a"/"year" confusion.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:162]: I think that reflects what Grahame has written. I guess this makes the unit test that he provided work?

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (ewoutkramer) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:178]: I don't see how this is fixed: adding a non-time unit to a date will, if I am correct, just turn the addition operation into a no-op (it wouldn't throw as there is no `default` case here).  I think the spec suggests we should return `{}` - that's what the existing other comparison operations do when incomparable units are used.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:154]: Excellent, I'll move that code there then (and test it out)

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:162]: Cool, I'll recheck that there are enough unit test cases covering all these combinations.
(existing ones pass)
And wait on an answer tomorrow on which way the fhir chat points it.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:178]: Yup, will rectify, I must have Somme Ted on the wrong one.

Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:183]: There is now a default case for the switch that throws, this should be resolved.

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:154]: If the ucum system is not defined in the FHIR quantity during the conversion, should it modify the unit value anyway?

Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs:154]: This code might get in our way... as you've specifically converted to the ucum variant.

https://github.com/FirelyTeam/firely-net-common/blob/3f94b54bc264f41cc0216ea7022935b60fd76626/src/Hl7.Fhir.ElementModel/Types/Quantity.cs#L152

And if I take the ucum cases out of the switches, others fail too
image

File: src/Hl7.Fhir.ElementModel/Types/Date.cs
Comment (brianpos) [reply src/Hl7.Fhir.ElementModel/Types/Date.cs]: All done now - thanks for helping get this done right...

**Comment by brianpos** [comment]
Have reported the 2 issues causing the failing unit tests on zulip
https://chat.fhir.org/#narrow/stream/179266-fhirpath/topic/date.20math
Hoping that we don't have to change anything further based on this - then we can release it.
(so wasn't quite as simple as my initial change was - even though it kinda worked right)

**Comment by brianpos** [comment]
And I didn't change the Implementation of the `typecasts.cs` FHIR.Questionnaire to the other native fhirpath ucum format - not sure if that's something you want covered before this is merged in.
(I'm ok with using it as is, and record that as an issue)

**Pull request merged**
